### ISSUE TITLE
security: bundle 7 low-risk, backwards-compatible security fixes

### ISSUE THREAD

**Issue by yingbull** [carlos-emr/carlos#1634]
## Summary

Seven low-risk, small-diff security fixes bundled into a single PR scope. All items are backwards-compatible: no API changes, no dependency bumps, no database migration, no integration-facing config changes. Each item closes one or more GitHub Security alerts identified during a multi-pass adversarial validation of the code-scanning results (April 2026).

**Total scope**: ~10 code-scanning alerts across 7 independent fixes. Estimated ~45 min of work for a senior developer; straightforward for any contributor familiar with the codebase.

---

## Fix 1 — Stored XSS in encounter navigation (`navigation.jsp`)

**File**: `src/main/webapp/casemgmt/navigation.jsp`
**Lines**: 405, 414

### Problem
`bean.userName` (originating from `EctSessionBean` populated via `EctIncomingEncounter2Action` / `EctDisplayAction`) is emitted raw into a JavaScript single-quoted string inside an HTML `onclick` attribute:
```jsp
onClick="popupPage('…&userName=<%=bean.userName%>')"
```
If a clinician's session state includes attacker-crafted characters (e.g. `' + alert(1) + '`), arbitrary JavaScript executes in every subsequent encounter page-load for that session. This is a genuine stored XSS with PHI-context impact.

### Fix
Wrap both emissions with the OWASP `Encode.forJavaScript()` helper already imported and used on line 323 of the same file:
```jsp
onClick="popupPage('…&userName=<%=Encode.forJavaScript(bean.userName)%>')"
```

### Backwards compatibility
- `Encode.forJavaScript` is already on the classpath and used in the same JSP — no new imports needed.
- Output is still a valid JavaScript string; benign user names (alphanumeric) pass through unchanged.
- No API/URL change.

### Acceptance criteria
- [ ] Line 405: wrap `bean.userName` with `Encode.forJavaScript()`
- [ ] Line 414: wrap `bean.userName` with `Encode.forJavaScript()`
- [ ] Manual smoke-test: open an encounter, navigate between messages — popupPage still works
- [ ] Manual XSS test: set a fake `userName` to `'+alert(1)+'` via session → no alert fires

### Related alerts
Closes GitHub code-scanning alerts on `EctIncomingEncounter2Action.java:283` and `EctDisplayAction.java:235` (the upstream session-write sites; consumer-side encoding is the correct fix).

---

## Fix 2 — Migrate signed-note integrity hash from MD5 to SHA-256 (`HashAudit`)

**File**: `src/main/java/io/github/carlos_emr/carlos/commn/model/HashAudit.java`
**Line**: 49

### Problem
`HashAudit` stores a hash fingerprint of signed clinical notes for tamper-audit purposes:
```java
private static final String ALGORITHM = "MD5";
```
MD5 collision resistance is cryptographically broken. While the `hash_audit` table is currently write-only (no verification path in the code today), the stored hash represents purported medico-legal integrity evidence. A regulator or forensics investigator relying on these records should have a modern algorithm.

### Fix
```java
private static final String ALGORITHM = "SHA-256";
```

### Backwards compatibility
- The `hash_audit` table stores hashes as a `VARCHAR` signature field; SHA-256 output is longer (64 hex chars vs MD5's 32). Verify the column length is sufficient:
  - MariaDB: `DESCRIBE hash_audit;` — check `signature` column width
  - If column is narrower than 64 chars, a small DDL bump is needed (still additive, no data loss)
- Existing MD5 hashes in the table remain valid historical records — new writes use SHA-256; any future verification routine should handle both via length detection.
- No code consumes the hash for authorization, so no runtime behavior changes.

### Acceptance criteria
- [ ] Change `ALGORITHM = "MD5"` → `"SHA-256"`
- [ ] Verify `hash_audit.signature` column can store 64 hex chars; add DDL migration if needed (new `database/mysql/updates/update-YYYY-MM-DD-hashaudit-sha256.sql`)
- [ ] Run existing `HashAuditTest` / regression suite to ensure write path still works
- [ ] No migration of existing rows required

### Related alerts
Closes the `java/InsecureHash` alert on `HashAudit.java:123` (and the corresponding Sonar `InsecureHash` duplicate).

---

## Fix 3 — Protocol-relative open redirect in `ReportReassign2Action`

**File**: `src/main/java/io/github/carlos_emr/carlos/mds/pageUtil/ReportReassign2Action.java`
**Line**: 256 (plus `newURL` construction at ~line 205)

### Problem
```java
String newURL = request.getRequestURI();
…
response.sendRedirect(newURL);
```
Tomcat preserves the path portion of the request line verbatim (minus CR/LF). A malicious request line like `GET //attacker.example/path HTTP/1.1` results in `getRequestURI()` returning `//attacker.example/path`. When passed to `response.sendRedirect(...)`, browsers interpret `//host/path` as a protocol-relative URL and redirect off-site. Classic open-redirect vector missed by the existing `forUriComponent` encoding of query params (which protects against query-portion injection, not path-portion authority smuggling).

### Fix (choose one)
Option A — explicit prefix (preferred, minimal change):
```java
String newURL = request.getContextPath() + request.getServletPath();
// then append query params (already OWASP-encoded)
```
Option B — guard check:
```java
if (newURL.startsWith("//") || newURL.startsWith("/\\")) {
    newURL = request.getContextPath() + "/";
}
```

### Backwards compatibility
- The current redirect resolves to a same-origin path; with Option A the user lands on the same URL structure as before for all legitimate requests.
- No behavior change for valid in-app navigation.

### Acceptance criteria
- [ ] Apply Option A (rebuild URL from `getContextPath()` + `getServletPath()`) OR Option B (explicit guard)
- [ ] Manual test: craft request `GET //evil.example/x HTTP/1.1` (e.g. via curl `--request-target`) → redirect stays same-origin
- [ ] Normal reassign flow still redirects to the correct servlet
- [ ] Consider extracting the `startsWith("//")` check into `RedirectValidationUtils` as a new helper if not already present

### Related alerts
Closes `javasecurity:S5146` (`java/unvalidated-url-redirection`) on line 256.

---

## Fix 4 — HTTP response splitting in `fieldnotereportdetail.jsp`

**File**: `src/main/webapp/eform/fieldNoteReport/fieldnotereportdetail.jsp`
**Line**: 53

### Problem
```jsp
String filename = residentName.replace(", ","").replace(" ","");
response.setHeader("Content-Disposition", "attachment; filename=" + filename);
```
`residentName` is a request parameter. The two `.replace(...)` calls only strip commas and spaces — **they do not strip CR/LF**. An attacker submitting `residentName=foo%0D%0ASet-Cookie:sid=evil` can inject arbitrary response headers.

### Fix
```jsp
String filename = residentName.replaceAll("[^A-Za-z0-9_\\-\\.]", "");
```
This whitelist approach allows only characters safe in filenames and header values. Alternatively explicitly strip control chars:
```jsp
String filename = residentName.replaceAll("[\\r\\n\\u0000-\\u001F\\u007F-\\u009F]", "").trim();
```

### Backwards compatibility
- Legitimate names (alphanumeric, hyphens, dots) pass through unchanged.
- Filenames with international characters get stripped — if this is a concern, add specific Unicode ranges to the whitelist (e.g., `[\p{L}\p{N}_\-.]`).
- No JS/HTML consumers depend on the non-stripped form.

### Acceptance criteria
- [ ] Replace both `.replace(...)` calls with a CR/LF-safe sanitizer (pattern above)
- [ ] Manual test: submit `residentName=AB%0D%0AX-Injected:1` → response has no injected header
- [ ] Normal PDF download still names the file correctly
- [ ] Consider factoring `sanitizeHeaderValue()` from `ManageDocument2Action.java` (already exists, lines ~1659) as a shared utility — if shared util pulled in, use that instead

### Related alerts
Closes `java/HttpResponseSplitting` on line 53.

---

## Fix 5 — Backslash bypass in `CaseManagementEntry2Action.isValidInternalRedirect`

**File**: `src/main/java/io/github/carlos_emr/carlos/casemgmt/web/CaseManagementEntry2Action.java`
**Line**: ~3855 (the local `isValidInternalRedirect` method)
**Caller line**: ~2028-2030

### Problem
The locally-defined validator only checks:
```java
if (url.startsWith("/") && !url.startsWith("//")) return true;
if (url.contains("://")) return false;
```
This **misses the backslash variants**: `/\evil.example`, `\\evil.example`, URL-encoded `%5C`. Browsers normalize `/\host` to `//host` at the network layer, routing off-site. Open redirect confirmed via adversarial trace.

A hardened utility already exists in the codebase: `io.github.carlos_emr.carlos.utility.RedirectValidationUtils.isValidRelativeRedirect` — it correctly rejects backslash, `%5C`, encoded `..`, protocol-relative, absolute URIs, and javascript:/data:/vbscript:.

### Fix
Replace the local validator with a call to the shared utility:
```java
// Delete the local isValidInternalRedirect(String) method
// Replace all callers:
if (RedirectValidationUtils.isValidRelativeRedirect(chain)) { … }
```

### Backwards compatibility
- The shared utility is strictly stricter than the local one, never looser. All inputs that pass the local check continue to pass.
- Existing internal redirect URLs (starting with `/web/...`, `/casemgmt/...`) remain valid.
- No API-surface change.

### Acceptance criteria
- [ ] Delete the local `isValidInternalRedirect` method from `CaseManagementEntry2Action.java`
- [ ] Update all internal callers to use `RedirectValidationUtils.isValidRelativeRedirect`
- [ ] Add import for `io.github.carlos_emr.carlos.utility.RedirectValidationUtils`
- [ ] Unit test: feed `/\evil.example`, `/%5Cevil.example`, `\\evil.example` → all should return `false`
- [ ] Manual smoke test: legitimate `chain=/web/encounter/…` redirects still work

### Related alerts
Closes `java/OR` (open redirect) on line 2030.

---

## Fix 6 — Bounded admin regex input on `updatedemographicprovider.jsp`

**File**: `src/main/webapp/admin/updatedemographicprovider.jsp`
**Line**: 244
**Downstream sink**: `DemographicManagerImpl.getDemographicsNameRangeByProvider` line ~380 (`Pattern.compile(regex)`)

### Problem
```jsp
String regexp = request.getParameter("regexp");
…
Pattern.compile(regexp);
```
Admin users can submit unbounded regex input, enabling ReDoS (regex denial-of-service). While admin-only, a stray long expression accidentally pasted could lock up a Tomcat worker thread.

### Fix
Before `Pattern.compile`, bound the length and reject pathological inputs:
```jsp
if (regexp == null || regexp.length() > 100) {
    // reject or truncate
    regexp = (regexp == null) ? "" : regexp.substring(0, 100);
}
```
Consider also wrapping the compile call with a short-circuit timeout if ReDoS-hardening is a broader concern — but length-bounding is the minimum needed to close the alert.

### Backwards compatibility
- 100 chars is well above typical admin-regex use (patient name prefix, department code).
- Null guard prevents NPE if parameter absent.
- No workflow change for normal admin usage.

### Acceptance criteria
- [ ] Add length-bound (≤100 chars) before `Pattern.compile`
- [ ] Handle null parameter defensively
- [ ] Manual test: admin page still filters demographics as expected
- [ ] Manual test: paste a 1000-char regex → handled gracefully (truncated or rejected, not hanging)

### Related alerts
Closes `java/RegexInjection` on line 244.

---

## Fix 7 — Missing admin gate on duplicate `obarchecklistedit_99_12.jsp` + `obarriskedit_99_12.jsp`

**Files**:
- `src/main/webapp/provider/obarchecklistedit_99_12.jsp` (unpatched duplicate of `decision/antenatal/obarchecklistedit_99_12.jsp`)
- `src/main/webapp/decision/antenatal/obarriskedit_99_12.jsp` (sibling with same issue)

### Problem
The hardened pattern (admin gate + secure SAX parser) was applied to `src/main/webapp/decision/antenatal/obarchecklistedit_99_12.jsp` but **not** propagated to:
1. The provider-path duplicate (`provider/obarchecklistedit_99_12.jsp`) — any authenticated user can POST `submit=" Save "` and overwrite an XML config file on disk via `FileWriter`.
2. `obarriskedit_99_12.jsp` in the same decision/antenatal directory — lacks admin gate, uses only weak whitespace-based XML escaping instead of a secure parser.

### Fix
Copy the hardening block from the already-fixed `decision/antenatal/obarchecklistedit_99_12.jsp`:
```jsp
<%-- at top of JSP, before any state-changing logic --%>
<%
SecurityInfoManager securityInfoManager = SpringUtils.getBean(SecurityInfoManager.class);
if (!securityInfoManager.hasPrivilege(LoggedInInfo.getLoggedInInfoFromSession(request), "_admin", "w", null)) {
    throw new SecurityException("missing required sec object");
}
%>
…
<%-- before writing XML --%>
// use XmlUtils.createSecureSAXParserFactory() to validate well-formedness
// before FileWriter
```

### Backwards compatibility
- Admin users with `_admin,w` continue to function identically.
- Non-admin access previously silently succeeded; now returns 500 (SecurityException). Identify any legitimate non-admin use case during review — if none, the closure is correct.
- If the `provider/` path is truly an orphaned duplicate, consider deleting it entirely and redirecting any references (grep confirms no other file references it).

### Acceptance criteria
- [ ] `provider/obarchecklistedit_99_12.jsp`: copy admin gate + secure SAX parser from `decision/antenatal/` sibling, OR delete the file if orphaned (grep for usages first)
- [ ] `decision/antenatal/obarriskedit_99_12.jsp`: add same admin gate; replace whitespace-based XML escaping with `XmlUtils.createSecureSAXParserFactory().newSAXParser()` validation before `FileWriter`
- [ ] Manual test: admin user can still save checklists/risks
- [ ] Manual test: non-admin POST returns 500 (no file write)
- [ ] Grep `src/main/webapp/` for any JSP/link pointing to `provider/obarchecklistedit_99_12.jsp` — none expected

### Related alerts
Closes `javasecurity:S6399` alert on `decision/antenatal/obarriskedit_99_12.jsp:71`. (The `provider/` duplicate was discovered during adversarial audit and not directly flagged by the scanner.)

---

## Overall acceptance criteria

- [ ] All 7 fixes applied as a single PR or as a linked set of small PRs (preference: one PR for easier review)
- [ ] No new dependencies added
- [ ] No database migrations required beyond Fix 2's optional column-width DDL
- [ ] Existing test suite (`make install --run-modern-tests`) passes
- [ ] Manual smoke test of affected flows: encounter navigation, chart unlock, admin fax config, demographic search, report reassign, PDF download from fieldnote
- [ ] `make install` completes cleanly; Tomcat deploys without error

## Test plan

- [ ] Run `make install --run-modern-tests` — existing tests pass
- [ ] Smoke: log in as clinician, navigate encounters, verify popupPage works (Fix 1)
- [ ] Smoke: sign a note, verify `hash_audit` row written with longer hash (Fix 2)
- [ ] Adversarial: `curl --request-target //evil/x /mds/reportReassign.do` → stays same-origin (Fix 3)
- [ ] Adversarial: submit `residentName=A%0D%0AX-Injected:1` → response has no injected header (Fix 4)
- [ ] Unit: `RedirectValidationUtils.isValidRelativeRedirect("/\\evil.example")` returns false (Fix 5)
- [ ] Admin: paste 1000-char regex into demographics provider filter → handled, server doesn't hang (Fix 6)
- [ ] Admin: save obar checklist/risk as admin → success; as non-admin → 500 (Fix 7)

## Dismissing related code-scanning alerts

After merge, re-run code scanning. Alerts on the affected files should auto-close. Any residual alert that remains after the fix should be triaged individually — do not bulk-dismiss.

## Out of scope (handle separately)

The following related findings are intentionally excluded because they require design decisions or protocol coordination:
- `EncryptionUtils` key management (needs KMS/secret-store decision)
- `CxfClientUtils` / `EdtClientBuilder` TrustAllManager (needs cert-pinning strategy for lab/MCEDT integrations)
- `LabUpload2Action.java:217` RSA PKCS1v1.5 Bleichenbacher (needs upstream lab-partner coordination)
- `RptDownloadCSVServlet` / `RptFormQuery` SQL injection (admin-only report module; larger refactor)
- `EDocUtil.java` path injection (needs broader `PathValidationUtils` retrofit)
- `TraceabilityReportProcessor.java` Java deserialization (needs file-format migration)
- `unlockAjax.jsp` and other newly-discovered AJAX encoding gaps (separate systematic sweep — scanner did not flag these)

---

🤖 Based on multi-pass adversarial validation of GitHub code-scanning alerts (April 2026). See companion dismissal rationale on each closed FP alert.

**Comment by yingbull** [comment]
Covered by umbrella tracker #1636 (adversarial audit findings). Closing sub-issue; work tracked at the umbrella level.
